### PR TITLE
os/arch/arm/src/imxrt : Add PIT Timer

### PR DESCRIPTION
**Pull request by NitishAmbastha**
Implement PIT timer with 4 channels 32-bit counters
Supports Chain mode for higher count
Support LifeTime counter

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [29b89e222da2839d742516ca01a47b1c6c5a2c12] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [29b89e222da2839d742516ca01a47b1c6c5a2c12] - Code Rule Check (C++) OK.

File: os/arch/arm/src/imxrt/imxrt_pit.c [hunk 1]
```diff
@@ -151,3 +216,381 @@ uint64_t imxrt_pit_getlifetimetimercount(PIT_Type *base)
...
+
+		active++;
+
+		tmrvdbg("PIT Timer Channel %d is started, callback %s\n", priv->channel, priv->callback == NULL ? "none" : "set");
+		return OK;
+	}
+
+	/* Return EBUSY to indicate that the timer was already running */
+
+	tmrdbg("Error!! PIT Timer %d was already running\n", priv->channel);
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/imxrt/imxrt_pit.c]: This is not error case, please use verbose or warning message

File: os/arch/arm/src/imxrt/imxrt_pit.c [hunk 2]
```diff
@@ -151,3 +216,381 @@ uint64_t imxrt_pit_getlifetimetimercount(PIT_Type *base)
...
+
+		imxrt_pit_stoptimer(priv->pit->base, priv->channel);
+		if (ch_status[priv->channel]) {
+			imxrt_pit_stoptimer(priv->pit->base, priv->channel-1);
+			imxrt_pit_settimerchainmode(priv->pit->base, priv->channel, false);
+			ch_status[priv->channel] = false;
+			busy_status[priv->channel-1] = false;
+		}
+
+		if (active == 1) {
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/imxrt/imxrt_pit.c]: That's why I mentioned one more lowerhalf to control each of channels. you are counting another dev's status here. All lower-half should work independently.

File: os/arch/arm/src/imxrt/imxrt_pit.c [hunk 3]
```diff
@@ -151,3 +216,381 @@ uint64_t imxrt_pit_getlifetimetimercount(PIT_Type *base)
...
+ *           command.
+ *
+ * Returned Values:
+ *   Zero on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+static int imxrt_pit_ioctl(struct timer_lowerhalf_s *lower, int cmd,
+						unsigned long arg)
+{
+	return OK;
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/imxrt/imxrt_pit.c]: -ENOTTY

File: os/arch/arm/src/imxrt/imxrt_pit.c [hunk 4]
```diff
@@ -151,3 +216,381 @@ uint64_t imxrt_pit_getlifetimetimercount(PIT_Type *base)
...
+	lower->started  = false;
+	lower->callback = NULL;
+
+	imxrt_pit_getdefaultconfig(&config);
+	memcpy(&(lower->config), &config, sizeof(pit_config_t));
+	lower->pit = imxrt_pit_configure(timer, &config, timer);
+	lower->channel = timer;
+
+	if (lower->pit == NULL) {
+		return -EINVAL;
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/imxrt/imxrt_pit.c]: You didn't use any of parameter here, It's not -EINVAL.
If you add pit_config_t and use it here, then EINVAL can be accepted.

File: os/arch/arm/src/imxrt/imxrt_pit.c [hunk 5]
```diff
@@ -151,3 +216,381 @@ uint64_t imxrt_pit_getlifetimetimercount(PIT_Type *base)
...
+
+	if (priv->started) {
+		imxrt_pit_disableinterrupts(priv->pit->base, priv->channel, kPIT_TimerInterruptEnable);
+
+		imxrt_pit_stoptimer(priv->pit->base, priv->channel);
+		if (ch_status[priv->channel]) {
+			imxrt_pit_stoptimer(priv->pit->base, priv->channel-1);
+			imxrt_pit_settimerchainmode(priv->pit->base, priv->channel, false);
+			ch_status[priv->channel] = false;
+			busy_status[priv->channel-1] = false;
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/imxrt/imxrt_pit.c]: if priv->channel is MXRT_PIT_CH0, then?

**Comment by NitishAmbastha** [comment]
Will re-design channels and update

**Comment by seinfra** [comment]
Target : [e6ac9cefc6570e04ceed330939388c9131b87ff8] - Code Rule Check Result:
os/arch/arm/src/imxrt/imxrt_pit.c:311: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/arch/arm/src/imxrt/imxrt_pit.c:318: ERROR: [SPC_M_KWD] space required before the open parenthesis '('

**Comment by seinfra** [comment]
Target : [e6ac9cefc6570e04ceed330939388c9131b87ff8] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [650049a7376041e311429a0ebdc5f28fe8079be0] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [650049a7376041e311429a0ebdc5f28fe8079be0] - Code Rule Check OK.

File: os/arch/arm/src/imxrt/imxrt_pit.c [hunk 6]
```diff
@@ -151,3 +217,373 @@ uint64_t imxrt_pit_getlifetimetimercount(PIT_Type *base)
...
+			imxrt_pit_starttimer(priv->pit->base, kPIT_Chnl_2);
+
+		tmrvdbg("ch_status:%d/%d/%d/%d\n", ch_status[0], ch_status[1], ch_status[2], ch_status[3]);
+#endif
+		tmrvdbg("PIT Timer started, callback %s\n", priv->callback == NULL ? "none" : "set");
+		return OK;
+	}
+
+	/* Return EBUSY to indicate that the timer was already running */
+
```
Comment (amandeepc) [inline os/arch/arm/src/imxrt/imxrt_pit.c]: We can remove these extra lines between comment and code

**Comment by seinfra** [comment]
Target : [2bf8770057613d996f77bebb051ef4d3779fe9df] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [2bf8770057613d996f77bebb051ef4d3779fe9df] - Code Rule Check Result:
os/arch/arm/src/imxrt/imxrt_pit.c:312: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
os/arch/arm/src/imxrt/imxrt_pit.c:319: ERROR: [SPC_M_KWD] space required before the open parenthesis '('

**Comment by seinfra** [comment]
Target : [ad1fbb65f8b8b3d02b7041f172ebdcf70c4d2641] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [3103617a968ad79986f7cd43b9faf0d44b4c7e06] - Code Rule Check OK.

File: os/arch/arm/src/imxrt/imxrt_pit.c (hunk 5 above)
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: Redesigned code

File: os/arch/arm/src/imxrt/imxrt_pit.c (hunk 4 above)
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: Updated

File: os/arch/arm/src/imxrt/imxrt_pit.c (hunk 3 above)
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: Implemented ioctl for lifetimer. No more required to return -ENOTTY

File: os/arch/arm/src/imxrt/imxrt_pit.c (hunk 2 above)
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: Done

File: os/arch/arm/src/imxrt/imxrt_pit.c (hunk 1 above)
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: Done

File: os/arch/arm/src/imxrt/imxrt_pit.c [hunk 7]
```diff
@@ -151,3 +216,381 @@ uint64_t imxrt_pit_getlifetimetimercount(PIT_Type *base)
...
+			uint32_t ch_ticks1 = (uint32_t)freq /USEC_PER_SEC;
+			uint32_t ch_ticks2 = (uint32_t)timeout;
+			tmrvdbg("ticks:%llu ch_ticks1:%d ch_ticks2:%d\n", ticks, ch_ticks1, ch_ticks2);
+			imxrt_pit_settimerchainmode(priv->pit->base, priv->channel, true);
+			imxrt_pit_settimerperiod(priv->pit->base, priv->channel, ch_ticks1);
+			imxrt_pit_settimerperiod(priv->pit->base, chained, ch_ticks2);
+			ch_status[priv->channel] = true;
+		} else {
+			tmrvdbg("CHAIN mode cannot be used as channel[%d] is Busy\n", chained);
+			return -EBUSY;
```
[the remark this answers is not in the document]
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: Done

**Comment by seinfra** [comment]
Target : [3103617a968ad79986f7cd43b9faf0d44b4c7e06] - Code Rule Check (C++) OK.

File: os/arch/arm/src/imxrt/imxrt_pit.c [hunk 8]
```diff
@@ -151,3 +216,381 @@ uint64_t imxrt_pit_getlifetimetimercount(PIT_Type *base)
...
+	}
+
+	freq = imxrt_clock_getperclkfreq();
+	ticks = ((uint64_t)freq * (uint64_t)timeout) / USEC_PER_SEC;
+
+	if (ticks < UINT_MAX) {
+		imxrt_pit_settimerperiod(priv->pit->base, priv->channel, ticks);
+	} else if (ticks >= UINT_MAX) {
+		pit_chnl_t chained = priv->channel - 1;
+		if (busy_status[chained] == false) {
```
[the remark this answers is not in the document]
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: Redesigned the code.

File: os/arch/arm/src/imxrt/imxrt_pit.c [hunk 9]
```diff
@@ -151,3 +216,381 @@ uint64_t imxrt_pit_getlifetimetimercount(PIT_Type *base)
...
+	struct imxrt_pit_lowerhalf_s *priv = (struct imxrt_pit_lowerhalf_s *)lower;
+
+	if (priv->started) {
+		return -EPERM;
+	}
+
+	freq = imxrt_clock_getperclkfreq();
+	ticks = ((uint64_t)freq * (uint64_t)timeout) / USEC_PER_SEC;
+
+	if (ticks < UINT_MAX) {
```
[the remark this answers is not in the document]
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: Done

File: os/arch/arm/src/imxrt/imxrt_pit.c (hunk 6 above)
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: Done

**Comment by seinfra** [comment]
Target : [6b546c79e22e32fbfb489e04fb3819c11d3336e1] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [cff53d8ea07a9cc519286643e5a6dfff9c5a7060] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [cff53d8ea07a9cc519286643e5a6dfff9c5a7060] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [27a0686f1c6a66623b161f55da17868d1a1d25f9] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [27a0686f1c6a66623b161f55da17868d1a1d25f9] - Code Rule Check (C++) OK.

File: os/arch/arm/src/imxrt/imxrt_pit.c [hunk 10]
```diff
@@ -1,153 +1,589 @@
...
+		imxrt_pit_settimerperiod(priv->pit->base, kPIT_Chnl_3, ticks);
+#if defined(FSL_FEATURE_PIT_HAS_CHAIN_MODE) && FSL_FEATURE_PIT_HAS_CHAIN_MODE
+	} else if (ticks >= UINT_MAX) {
+		/* Chain mode supported on channel 2 and 3 only */
+		tmrvdbg("Enter CHAIN mode\n");
+		uint32_t ch_ticks1 = (uint32_t)freq /USEC_PER_SEC;
```
Comment (sunghan-chang) [inline os/arch/arm/src/imxrt/imxrt_pit.c]: `/ USEC`

File: os/arch/arm/src/imxrt/imxrt_pit.c [hunk 11]
```diff
@@ -1,153 +1,589 @@
...
+	imxrt_pit_clearstatusflags(pit->base, kPIT_Chnl_3, kPIT_TimerFlag);
+	if (ch_status[kPIT_Chnl_3])
+		imxrt_pit_clearstatusflags(pit->base, kPIT_Chnl_2, kPIT_TimerFlag);
+
+	if (priv->callback(&next_interval_us, priv->arg)) {
+		if (next_interval_us > 0) {
+			freq = imxrt_clock_getperclkfreq();
+			ticks = ((uint64_t)freq * (uint64_t)next_interval_us) / USEC_PER_SEC;
+			if (ch_status[kPIT_Chnl_3]) {
+				uint32_t ch_ticks1 = (uint32_t)freq /USEC_PER_SEC;
```
Comment (sunghan-chang) [inline os/arch/arm/src/imxrt/imxrt_pit.c]: same as above

File: os/arch/arm/src/imxrt/imxrt_pit.h [hunk 12]
```diff
@@ -26,6 +26,7 @@
 #define _IMXRT_PIT_H_
 
 #include "imxrt_config.h"
+#include <chip/imxrt105x_features.h>
```
[the remark this answers is not in the document]
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.h]: Most of the features of PIT like PIT timer count, Lifetime timer, Chaining mode and MDIS settings are defined in imxrt105x_features. This file needs to be included to enable them.

File: os/board/imxrt1050-evk/src/imxrt_boot.c [hunk 13]
```diff
@@ -278,14 +281,18 @@ void board_initialize(void)
...
 
 #ifdef CONFIG_IMXRT_GPT
 		for (timer_idx = 0; timer_idx < IMXRT_GPT_CH_MAX; timer_idx++) {
 			snprintf(timer_path, sizeof(timer_path), "/dev/timer%d", timer_idx);
 			imxrt_timer_initialize(timer_path, timer_idx);
 		}
+#endif
+#ifdef CONFIG_IMXRT_PIT
+		snprintf(timer_path, sizeof(timer_path), "/dev/timer%d", timer_idx);
+		imxrt_pit_initialize(timer_path);
```
[the remark this answers is not in the document]
Comment (NitishAmbastha) [reply os/board/imxrt1050-evk/src/imxrt_boot.c]: Yes, there is 1 PIT on IMXRT. It has 4 channels which are managed from lower half internally.
Will push changes for imxrt1020-evk in sometime

File: os/arch/arm/src/imxrt/imxrt_pit.c (hunk 10 above)
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: ok

File: os/arch/arm/src/imxrt/imxrt_pit.c (hunk 11 above)
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: ok

**Comment by seinfra** [comment]
Target : [6a0ee49a8a7c4d2827968533c66e9c05c074b17a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [6a0ee49a8a7c4d2827968533c66e9c05c074b17a] - Code Rule Check OK.

File: os/arch/arm/src/imxrt/imxrt_pit.c [hunk 14]
```diff
@@ -1,153 +1,605 @@
...
+ * The period of lifetime timer is equal to the "period of timer 0 * period of timer 1".
+ *
+ * param base PIT peripheral base address
+ */
+static inline void imxrt_pit_start_lifetimetimercount(PIT_Type *base)
+{
+	tmrvdbg("Start lifetime counter\n");
+	base->CHANNEL[kPIT_Chnl_1].LDVAL = 0xFFFFFFFF;
+	base->CHANNEL[kPIT_Chnl_1].TCTRL = 0x0;
+	base->CHANNEL[kPIT_Chnl_1].TCTRL |= PIT_TCTRL_CHN_MASK;
```
Comment (sunghan-chang) [inline os/arch/arm/src/imxrt/imxrt_pit.c]: Need to include `chip/imxrt102x_config.h` and `chip/imxrt105x_config.h`

File: os/arch/arm/src/imxrt/imxrt_pit.h [hunk 15]
```diff
@@ -1,342 +1,381 @@
...
+ * Copyright 2016-2017 NXP
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#ifndef _IMXRT_PIT_H_
+#define _IMXRT_PIT_H_
+
+#include "imxrt_config.h"
+#include <chip/imxrt105x_features.h>
```
Comment (sunghan-chang) [inline os/arch/arm/src/imxrt/imxrt_pit.h]: You should include chip/imxrt102x_features.h as well.

Comment (sunghan-chang) [reply os/arch/arm/src/imxrt/imxrt_pit.h]: #if defined(CONFIG_ARCH_CHIP_FAMILY_IMXRT102x)
#include "chip/imxrt102x_config.h"
#elif defined(CONFIG_ARCH_CHIP_FAMILY_IMXRT105x)
#include "chip/imxrt105x_config.h"
#endif

File: os/board/imxrt1050-evk/src/imxrt_boot.c [hunk 16]
```diff
@@ -290,14 +293,18 @@ void board_initialize(void)
...
 		char timer_path[CONFIG_PATH_MAX];
 
 #ifdef CONFIG_IMXRT_GPT
 		for (timer_idx = 0; timer_idx < IMXRT_GPT_CH_MAX; timer_idx++) {
 			snprintf(timer_path, sizeof(timer_path), "/dev/timer%d", timer_idx);
 			imxrt_timer_initialize(timer_path, timer_idx);
 		}
+#endif
+#ifdef CONFIG_IMXRT_PIT
+		snprintf(timer_path, sizeof(timer_path), "/dev/timer%d", timer_idx);
```
Comment (sunghan-chang) [inline os/board/imxrt1050-evk/src/imxrt_boot.c]: 1.  If it has only one pit, we don't need to use this snprintf.
2. As qtmr does, let's use different name.
timer -> pit

File: os/board/imxrt1050-evk/src/imxrt_boot.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (sunghan-chang) [reply os/board/imxrt1050-evk/src/imxrt_boot.c]: As I notice, currently 1020 is used, not 1050. I can merge this if it is in 1020 board. But I can't merge only with 1050 added.

File: os/arch/arm/src/imxrt/imxrt_pit.c (hunk 14 above)
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: Not needed. This is been included in ./arch/arm/src/imxrt/imxrt_clock.h

File: os/arch/arm/src/imxrt/imxrt_pit.h (hunk 15 above)
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.h]: ok

File: os/arch/arm/src/imxrt/imxrt_pit.c (hunk 14 above)
Comment (NitishAmbastha) [reply os/arch/arm/src/imxrt/imxrt_pit.c]: ok

**Comment by seinfra** [comment]
Target : [46dd37de9372b04e82b66f5dfcba30ad30ecc26e] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [46dd37de9372b04e82b66f5dfcba30ad30ecc26e] - Code Rule Check OK.

File: os/board/imxrt1050-evk/src/imxrt_boot.c (hunk 16 above)
Comment (NitishAmbastha) [reply os/board/imxrt1050-evk/src/imxrt_boot.c]: ok done.

**Comment by NitishAmbastha** [comment]
Changes pushed

**Comment by seinfra** [comment]
Target : [c2b8f8221f62cc2851c4e75cf00e88667efef837] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [c2b8f8221f62cc2851c4e75cf00e88667efef837] - Code Rule Check OK.

File: os/board/imxrt1050-evk/src/imxrt_boot.c [hunk 17]
```diff
@@ -86,7 +86,10 @@
 #include "imxrt_adc.h"
 #endif
 
-
+#ifdef CONFIG_IMXRT_PIT
+#include "imxrt_pit.h"
+#define PIT_DEVPATH     "/dev/timer-pit"
```
Comment (sunghan-chang) [inline os/board/imxrt1050-evk/src/imxrt_boot.c]: I think "/dev/pit" or "/dev/pitimer" is better.

**Comment by seinfra** [comment]
Target : [d78b3a20f3186689578a05dd5d610054a2553706] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d78b3a20f3186689578a05dd5d610054a2553706] - Code Rule Check OK.

**Comment by sunghan-chang** [comment]
@NitishAmbastha Please resolve the conflict.

**Comment by seinfra** [comment]
Target : [5fabb448cacef4324da43615606667abc95d366f] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [5fabb448cacef4324da43615606667abc95d366f] - Code Rule Check (C++) OK.

**Pull request merged**